scripts/packages/VSCodeServer/src/eval.jl

```diff
@@ -131,3 +95,69 @@ function safe_render(x)
         )
     end
 end
+
+"""
+    render(x)
+
+Produce a representation of `x` that can be displayed by a UI.
+Must return a `ReplRunCodeRequestReturn` with the following fields:
+- `inline::String`: Short one-line plain text representation of `x`. Typically limited to `INLINE_RESULT_LENGTH` characters.
+- `all::String`: Plain text string (that may contain linebreaks and other signficant whitespace) to further describe `x`.
+- `stackframe::Vector{Frame}`: Optional, should only be given on an error
+"""
+function render(x)
+    str = sprintlimited(MIME"text/plain"(), x, limit=MAX_RESULT_LENGTH)
+    inline = strlimit(first(split(str, "\n")), limit=INLINE_RESULT_LENGTH)
+    all = codeblock(str)
+    return ReplRunCodeRequestReturn(inline, all)
+end
+
+render(::Nothing) = ReplRunCodeRequestReturn("✓", codeblock("nothing"))
+
+codeblock(s) = string("```julia", '\n', s, '\n', "```")
+
+struct EvalError
+    err
+    bt
+end
+
+function render(err::EvalError)
+    bt = err.bt
+    i = find_frame_index(bt, @__FILE__, inlineeval)
+    bt = bt[1:(i === nothing ? end : i - 4)]
+    st = stacktrace(bt)
+
+    errstr = sprintlimited(err.err, nothing, func = Base.display_error, limit = MAX_RESULT_LENGTH)
+    inline = strlimit(first(split(errstr, "\n")), limit = INLINE_RESULT_LENGTH)
+    all = string(codeblock(errstr), '\n', backtrace_string(bt))
+    stackframe = Frame.(st)
+    return ReplRunCodeRequestReturn(inline, all, stackframe)
+end
+
+# more cleaner way ?
+const LOCATION_REGEX = r"\[\d+\]\s(?<body>.+)\sat\s(?<path>.+)\:(?<line>\d+)"
+
+function backtrace_string(bt)
+    s = sprintlimited(bt, func = Base.show_backtrace, limit = MAX_RESULT_LENGTH)
+    lines = strip.(split(s, '\n'))
+
+    return map(enumerate(lines)) do (i, line)
+        i === 1 && return line # "Stacktrace:"
+        m = match(LOCATION_REGEX, line)
+        m === nothing && return line
+        linktext = string(m[:path], ':', m[:line])
+        args = JSON.json((path = fullpath(m[:path]), line = m[:line]))
+        linkbody = vscode_cmd_string("language-julia.openFile", args)
+        linktitle = string("Go to ", linktext)
+        return "$(i-1). `$(m[:body])` at [$(linktext)]($(linkbody) \"$(linktitle)\")"
+    end |> joinlines
+end
```

Comment:
I couldn't come up with any good idea to implement this, other than using ugly `Regex` ...
I think we can also hack into the `Base` internals, but I don't think it would look clean.